src/btree/bt_vrfy.c

```diff
@@ -594,84 +588,98 @@ __verify_addr_ts(WT_SESSION_IMPL *session, WT_REF *ref, WT_CELL_UNPACK *unpack,
 
-/*
- * __verify_btree_id_with_meta --
- *     With a given btree id, iterate through the metafile table and check if the btree id exists in
- *     any table. If exists, we place the table URI into a variable otherwise should return
- *     WT_NOTFOUND.
- */
-static int
-__verify_btree_id_with_meta(WT_SESSION_IMPL *session, uint32_t btree_id, const char **uri)
-{
-    WT_CONFIG_ITEM id;
-    WT_CURSOR *meta_cursor;
-    WT_DECL_RET;
-    const char *meta_value;
-
-    meta_cursor = NULL;
-    WT_RET(__wt_metadata_cursor(session, &meta_cursor));
-    while ((ret = meta_cursor->next(meta_cursor)) == 0) {
-        WT_ERR(meta_cursor->get_value(meta_cursor, &meta_value));
-        if ((ret = __wt_config_getones(session, meta_value, "id", &id)) == 0 &&
-          btree_id == id.val) {
-            WT_ERR(meta_cursor->get_key(meta_cursor, uri));
-            break;
-        }
-        WT_ERR_NOTFOUND_OK(ret);
-    }
-err:
-    WT_TRET(__wt_metadata_cursor_release(session, &meta_cursor));
-    return (ret);
-}
-
 /*
  * __wt_verify_history_store_tree --
- *     Verify the history store tree, by iterating and checking all the keys in the history and that
- *     it exists in both the meta file and the respective data store table. Skips duplicate keys by
- *     storing the previous key.
+ *     Verify the history store. There can't be an entry in the history store without having the
+ *     latest value for the respective key in the data store. If given a uri, limit the verification
+ *     to the corresponding btree.
  */
 int
-__wt_verify_history_store_tree(WT_SESSION_IMPL *session)
+__wt_verify_history_store_tree(WT_SESSION_IMPL *session, const char *uri)
 {
     WT_CURSOR *cursor, *data_cursor;
     WT_DECL_ITEM(tmp);
     WT_DECL_RET;
     WT_ITEM hs_key, prev_hs_key;
     WT_TIME_PAIR hs_start, hs_stop;
-    uint32_t btree_id, session_flags, prev_btree_id;
-    int cmp;
-    const char *uri;
+    uint32_t btree_id, btree_id_given_uri, session_flags, prev_btree_id;
+    int exact, cmp;
+    char *uri_itr;
 
     session_flags = 0;
-    prev_btree_id = 0; /* [-Wconditional-uninitialized] */
     data_cursor = NULL;
-    uri = NULL;
     WT_CLEAR(prev_hs_key);
     WT_CLEAR(hs_key);
+    btree_id_given_uri = 0; /* [-Wconditional-uninitialized] */
+    prev_btree_id = 0;      /* [-Wconditional-uninitialized] */
+    uri_itr = NULL;
 
     WT_ERR(__wt_scr_alloc(session, 0, &tmp));
     WT_ERR(__wt_hs_cursor(session, &session_flags));
     cursor = session->hs_cursor;
 
-    while ((ret = cursor->next(cursor)) == 0) {
+    /*
+     * If a uri has been provided, limit verification to the corresponding btree by jumping to the
+     * first record for that btree in the history store. Otherwise scan the whole history store.
+     */
+    if (uri != NULL) {
+        ret = __wt_metadata_uri_to_btree_id(session, uri, &btree_id_given_uri);
+        if (ret != 0)
+            WT_ERR_MSG(session, ret, "Unable to locate the URI %s in the metadata file.", uri);
```

Comment:
Don't use trailing periods in error messages, we stitch error messages together, so you can end up with random periods in random  places.